File: webapp/scripts/sync-merge.test.ts
// Run with: npm test  (Vitest). Assertions use node:assert/strict, which Vitest supports.
import { test } from 'vitest';
import assert from 'node:assert/strict';
import { mergeBundles, type SyncBundle } from '../src/domain/syncMerge.ts';

const now = 1_700_000_000_000;
const rec = (id: string, updatedAt: number, extra: Record<string, unknown> = {}) => ({
  id, type: 'txn', updatedAt, value: { id, updatedAt, ...extra },
});
const tomb = (id: string, deletedAt: number) => ({ id, type: 'txn', deletedAt });
const empty: SyncBundle = { records: [], tombstones: [] };

test('new record on remote is pulled in', () => {
  const r = mergeBundles(empty, { records: [rec('a', now - 100)], tombstones: [] }, now);
  assert.equal(r.upserts.length, 1);
  assert.equal(r.upserts[0].id, 'a');
  assert.equal(r.deletes.length, 0);
});

test('local-only record is untouched (no upsert needed)', () => {
  const r = mergeBundles({ records: [rec('a', now - 100)], tombstones: [] }, empty, now);
  assert.equal(r.upserts.length, 0);
  assert.equal(r.deletes.length, 0);
});

test('both edited same record → later updatedAt wins', () => {
  const local = { records: [rec('a', now - 200, { amt: 'L' })], tombstones: [] };
  const remote = { records: [rec('a', now - 100, { amt: 'R' })], tombstones: [] };
  const r = mergeBundles(local, remote, now);
  assert.equal(r.upserts.length, 1);
  assert.equal((r.upserts[0].value as { amt: string }).amt, 'R');

  // symmetric: if local is newer, nothing to apply locally
  const r2 = mergeBundles(remote, local, now);
  assert.equal(r2.upserts.length, 0);
});

test('remote delete newer than local edit → record removed locally', () => {
  const local = { records: [rec('a', now - 200)], tombstones: [] };
  const remote = { records: [], tombstones: [tomb('a', now - 100)] };
  const r = mergeBundles(local, remote, now);
  assert.equal(r.deletes.length, 1);
  assert.equal(r.deletes[0].id, 'a');
  assert.equal(r.upserts.length, 0);
  assert.ok(r.tombstones.some((t) => t.id === 'a'));
});

test('local edit newer than remote delete → record survives', () => {
  const local = { records: [rec('a', now - 50)], tombstones: [] };
  const remote = { records: [], tombstones: [tomb('a', now - 100)] };
  const r = mergeBundles(local, remote, now);
  assert.equal(r.deletes.length, 0);
  assert.equal(r.upserts.length, 0); // local already has the winning record
});

test('tie on timestamp → delete wins', () => {
  const local = { records: [rec('a', now - 100)], tombstones: [] };
  const remote = { records: [], tombstones: [tomb('a', now - 100)] };
  const r = mergeBundles(local, remote, now);
  assert.equal(r.deletes.length, 1);
});

test('old tombstones are GC-ed past the TTL', () => {
  const remote = { records: [], tombstones: [tomb('a', now - 200 * 24 * 3600 * 1000)] };
  const r = mergeBundles(empty, remote, now);
  assert.equal(r.tombstones.length, 0);
});

test('re-sync is idempotent (converged state yields no changes)', () => {
  const state = {
    records: [rec('a', now - 100), rec('b', now - 150)],
    tombstones: [tomb('c', now - 120)],
  };
  const r = mergeBundles(state, state, now);
  assert.equal(r.upserts.length, 0);
  assert.equal(r.deletes.length, 0);
});

File: test/unit/sync_merge_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:khazana/domain/services/sync_merge.dart';

const now = 1700000000000;

SyncRecord rec(String id, int updatedAt, [Map<String, dynamic> extra = const {}]) =>
    SyncRecord(id: id, type: 'txn', updatedAt: updatedAt, value: {'id': id, ...extra});
Tombstone tomb(String id, int deletedAt) =>
    Tombstone(id: id, type: 'txn', deletedAt: deletedAt);
const empty = SyncBundle();

void main() {
  const m = SyncMerge();

  test('new record on remote is pulled in', () {
    final r = m.merge(empty, SyncBundle(records: [rec('a', now - 100)]), now: now);
    expect(r.upserts.length, 1);
    expect(r.upserts.first.id, 'a');
    expect(r.deletes, isEmpty);
  });

  test('local-only record is untouched', () {
    final r = m.merge(SyncBundle(records: [rec('a', now - 100)]), empty, now: now);
    expect(r.upserts, isEmpty);
    expect(r.deletes, isEmpty);
  });

  test('both edited same record → later updatedAt wins', () {
    final local = SyncBundle(records: [rec('a', now - 200, {'amt': 'L'})]);
    final remote = SyncBundle(records: [rec('a', now - 100, {'amt': 'R'})]);
    final r = m.merge(local, remote, now: now);
    expect(r.upserts.length, 1);
    expect(r.upserts.first.value['amt'], 'R');
    expect(m.merge(remote, local, now: now).upserts, isEmpty);
  });

  test('remote delete newer than local edit → removed locally', () {
    final local = SyncBundle(records: [rec('a', now - 200)]);
    final remote = SyncBundle(tombstones: [tomb('a', now - 100)]);
    final r = m.merge(local, remote, now: now);
    expect(r.deletes.length, 1);
    expect(r.upserts, isEmpty);
    expect(r.tombstones.any((t) => t.id == 'a'), isTrue);
  });

  test('local edit newer than remote delete → survives', () {
    final local = SyncBundle(records: [rec('a', now - 50)]);
    final remote = SyncBundle(tombstones: [tomb('a', now - 100)]);
    final r = m.merge(local, remote, now: now);
    expect(r.deletes, isEmpty);
    expect(r.upserts, isEmpty);
  });

  test('tie → delete wins', () {
    final local = SyncBundle(records: [rec('a', now - 100)]);
    final remote = SyncBundle(tombstones: [tomb('a', now - 100)]);
    expect(m.merge(local, remote, now: now).deletes.length, 1);
  });

  test('old tombstones GC-ed past TTL', () {
    final remote = SyncBundle(tombstones: [tomb('a', now - 200 * 24 * 3600 * 1000)]);
    expect(m.merge(empty, remote, now: now).tombstones, isEmpty);
  });

  test('re-sync is idempotent', () {
    final state = SyncBundle(
      records: [rec('a', now - 100), rec('b', now - 150)],
      tombstones: [tomb('c', now - 120)],
    );
    final r = m.merge(state, state, now: now);
    expect(r.upserts, isEmpty);
    expect(r.deletes, isEmpty);
  });
}
